homeassistant/components/homekit/util.py

```diff
@@ -34,10 +40,47 @@ def validate_entity_config(values):
             code = config.get(ATTR_CODE)
             params[ATTR_CODE] = cv.string(code) if code else None
 
+        if domain == 'media_player':
+            mode = config.get(CONF_MODE)
+            if mode:
+                params[CONF_MODE] = cv.ensure_list(mode)
+                for mode in params[CONF_MODE]:
+                    if mode not in MEDIA_PLAYER_MODES:
+                        raise vol.Invalid(
+                            'Invalid mode: "{}", valid modes are: "{}".'
+                            .format(mode, MEDIA_PLAYER_MODES))
+
         entities[entity] = params
     return entities
 
 
+def validate_media_player_modes(state, config):
+    """Validate modes for media playeres."""
+    features = state.attributes.get(ATTR_SUPPORTED_FEATURES, 0)
+
+    supported_modes = []
+    if features & (SUPPORT_TURN_ON | SUPPORT_TURN_OFF):
+        supported_modes.append(ON_OFF)
+    if features & (SUPPORT_PLAY | SUPPORT_PAUSE):
+        supported_modes.append(PLAY_PAUSE)
+    if features & (SUPPORT_PLAY | SUPPORT_STOP):
+        supported_modes.append(PLAY_STOP)
+    if features & SUPPORT_VOLUME_MUTE:
+        supported_modes.append(TOGGLE_MUTE)
+
+    config_modes = config.get(CONF_MODE, MEDIA_PLAYER_MODES)
+
+    validated_modes = []
+    for mode in config_modes:
+        if mode in supported_modes and mode not in validated_modes:
+            validated_modes.append(mode)
+        if mode not in supported_modes:
+            _LOGGER.warning('The entity "%s" does not support '
```

Comment:
Since you only check if a `mode` is in the `list`, it isn't necessary to filter out duplicates with `and mode not in validated_modes`. Through the change in `vec` we guaranty that `config[CONF_MODE]` is always a `list` at the time of execution for this function, so we can directly iterate over it. Furthermore I think we should raise `vol.Invalid` if a mode is not supported.
```py
if not config[CONF_MODE]:
    config[CONF_MODE] = supported_modes
    return

for mode in config[CONF_MODE]:
    if mode not in supported_modes:
        raise vol.Invalid('......')
```

Reply:
With this in `validate_entity_config`
```python
        if domain == 'media_player':
            mode = config.get(CONF_MODE)
            params[CONF_MODE] = cv.ensure_list(mode)
            for key in params[CONF_MODE]:
                if key not in MEDIA_PLAYER_MODES:
                    raise vol.Invalid(
                        'Invalid mode: "{}", valid modes are: "{}".'
                        .format(key, MEDIA_PLAYER_MODES))
```
And this in `validate_media_player_modes`
```python
    if not config[CONF_MODE]:
        config[CONF_MODE] = supported_modes
        return

    for mode in config[CONF_MODE]:
        if mode not in supported_modes:
            raise vol.Invalid('"{}" does not support mode: "{}".'
                              .format(state.entity_id, mode))
```
An error is produced if `mode` is not given in `entity_config`
`KeyError: 'mode'`
I know that changing to `if not config.get(CONF_MODE):` works but is there a better solution?

Reply:
I couldn't reproduce the error. Can you push your changes to Github, so I can try them?